Introduce outputVariables in StatisticAggregations

kewang1024 opened the pull request:
Context:
In the query plan for table writer, Coordinator sends StatisticAggregations
to indicate what column statistics we need to collect, for example:
name->count
name -> approx_distinct
name -> sum_data_size_for_stats
name -> max_data_size_for_stats
age -> min
age -> max
age -> count_8
The order should be fixed and it's used to determine output columns.
<img width="871" alt="image" src="https://github.com/prestodb/presto/assets/50636602/ba751381-2dd7-4802-82de-4fb28f857d61">

In java coordinator, this data structure is stored in RegularImmutableMap
and it's an ordered map based on its insertion order.

In java worker, after being deserialized back to RegularImmutableMap,
it's able to preserve its insertion order.
However in cpp worker, after being deserialized to std::map, it no longer
holds the insertion order.

To make cpp worker aware of the order of aggregations in the output,
we introduce outputVariables in StatisticAggregations.
 


```
== NO RELEASE NOTE ==
```

xiaoxmeng commented on presto-main/src/main/java/com/facebook/presto/sql/planner/plan/StatisticAggregations.java:
```diff
@@ -25,31 +25,51 @@
+        aggregations = aggregationMapBuilder.build();
+    }
+
+    public StatisticAggregations(
```
nit: Is this still required? Thanks!

kewang1024 replied:
Yes, basically in java code base, they're still heavily rely on the map structure, so if we only keep the List, the lookup would not be efficient.

So essentially we only introduced the List structure for serialization and deserialization, internally the logic remained the same.

xiaoxmeng reviewed (approved):
@kewang1024 LGTM. Please also have @mbasmanova to take a look as well? Thanks!

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/sql/planner/plan/StatisticAggregations.java:
```diff
@@ -25,31 +25,51 @@
 {
+    // Added to preserve the order for serializing and deserializing
+    // map in c++ worker doesn't represent insertion order
+    private final List<VariableAggregation> aggregationList;
```
> VariableAggregation

This name it confusing. It reads as "variable aggregation", not as a "{variable, aggregation} pair". I wonder if there is a better name.

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/sql/planner/plan/StatisticAggregations.java:
```diff
@@ -25,31 +25,51 @@
     @JsonCreator
     public StatisticAggregations(
-            @JsonProperty("aggregations") Map<VariableReferenceExpression, Aggregation> aggregations,
+            @JsonProperty("aggregationList") List<VariableAggregation> aggregationList,
```
aggregationList -> aggregations for consistency

Note that groupingVariables is also a List, but doesn't have "List" suffix.

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/sql/planner/plan/StatisticAggregations.java:
```diff
@@ -147,4 +172,32 @@ public StatisticAggregations getPartialAggregation()
+    @Immutable
+    public static class VariableAggregation
+    {
+        private final VariableReferenceExpression variableReferenceExpression;
```
I wonder if it would be cleaner to maintain 2 parallel lists instead of a list of these structs.

List<VariableReferenceExpression> aggregateVariables;
List<Aggregation> aggregations;

mbasmanova reviewed (commented):
Some comments.

mbasmanova commented on presto-native-execution/presto_cpp/presto_protocol/special/TableWriterNode.cpp.inc:
```diff
@@ -16,15 +16,51 @@ TableWriterNode::TableWriterNode() noexcept {
+
+}
+
+void from_json(const json& j, VariableAggregation& p) {
```
Do we need these customizations? Why? What happens if we remove these files?

kewang1024 replied:
Yes, because there are still column that is not included in prestissimo, for example `@JsonProperty("target") Optional<WriterTarget> target`
Inside `WriterTarget`, it has more data structures that we don't support now, my guess that `TableWriterNode.cpp.inc` existed in the first place is that we don't have a full solution for this class yet and we can remove it once we add the last piece

kewang1024 commented:
@mbasmanova I have adjusted the structure and updated the description accordingly 

FYI, I removed the change on prestissimo for now, we can first agree on java side's change and then I can add prestissimo's change in next step

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/sql/planner/plan/StatisticAggregations.java:
```diff
@@ -25,31 +25,70 @@
+    // map in c++ worker doesn't represent insertion order.
+    private final List<VariableReferenceExpression> variables;
+    private final List<Aggregation> variableAggregations;
+    // Only used internally for quick lookup.
```
This comment suggests that this map is used internally in this class and is not exposed externally. However, I see getAggregations() method returning this value directly. Hence, this comment is incorrect. Looking at the usages of getAggregations(), I see that these are used in aggregations and I also see that AggregationNode uses a map to specify the order of aggregations in the output. This means we have the same problem with regular aggregations as with aggregations used to collect stats. 

The issue with regular aggregations gets masked by Velox using column names instead of channels (indices) to lookup input columns in the operators. 


```
    @JsonCreator
    public AggregationNode(
            Optional<SourceLocation> sourceLocation,
            @JsonProperty("id") PlanNodeId id,
            @JsonProperty("source") PlanNode source,
            @JsonProperty("aggregations") Map<VariableReferenceExpression, Aggregation> aggregations,
            @JsonProperty("groupingSets") GroupingSetDescriptor groupingSets,
            @JsonProperty("preGroupedVariables") List<VariableReferenceExpression> preGroupedVariables,
            @JsonProperty("step") Step step,
            @JsonProperty("hashVariable") Optional<VariableReferenceExpression> hashVariable,
            @JsonProperty("groupIdVariable") Optional<VariableReferenceExpression> groupIdVariable)
```

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/sql/planner/plan/StatisticAggregations.java:
```diff
@@ -25,31 +25,70 @@
+    // Added to preserve the order for serializing and deserializing.
+    // map in c++ worker doesn't represent insertion order.
+    private final List<VariableReferenceExpression> variables;
+    private final List<Aggregation> variableAggregations;
```
'variableAggregations' name doesn't match the value; why 'variable'?

I see that name 'aggregations' is used by the Map variable. I wonder whether we need Map variable at all? Quickly checking some usages, it looks like Map is not needed and a list would work just fine. If we do need it, maybe we can rename the Map variable to variableToAggregationMap or similar.

Alternatively, maybe we can drop the new List<Aggregation> variable and use just List<VariableReferenceExpression> + the original map. This should work and may be less hacky.

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/sql/planner/plan/StatisticAggregations.java:
```diff
@@ -25,31 +25,70 @@
 
     @JsonProperty
-    public Map<VariableReferenceExpression, Aggregation> getAggregations()
+    public List<VariableReferenceExpression> getVariables()
```
Let's document this variable to explain that it specifies the order of aggregations in the output. In other places, I saw this called outputVariables. Should we used the same name here?

mbasmanova reviewed (commented):
@kewang1024 Ke, thank you for iterating on this PR. Some further comments.

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/sql/planner/plan/StatisticAggregations.java:
```diff
@@ -34,18 +34,37 @@
             @JsonProperty("aggregations") Map<VariableReferenceExpression, Aggregation> aggregations,
             @JsonProperty("groupingVariables") List<VariableReferenceExpression> groupingVariables)
     {
+        this.outputVariables = ImmutableList.copyOf(requireNonNull(outputVariables, "outputVariables is null"));
```
nit: perhaps, check that outputVariables matches the keys of aggregations

mbasmanova reviewed (approved):
@kewang1024 Thanks.
